### ISSUE TITLE
SYS registers need to be in a fixed location.

### ISSUE THREAD

**Issue by paulscottrobson** [commanderx16/x16-rom#62]
SYS uses three memory locations - sareg, sxreg,syreg and spreg to pass A,X,Y,S in and out of machine code routines. These need to have specified locations (they are 780-783 in the C64) as it will be somewhat challenging if they move about between ROM builds.

**Comment by greg-king5** [comment]
I prefer that the X16 implements the C128/Plus4 versions of the SYS and RREG statements:
```bas
SYS PC, A, X, Y, PSW
RREG AV, XV, YV, PV
```

**Comment by EgonOlsen71** [comment]
I was under the impression that it's the same on the X16. At least up to r33, they seem to be.

**Comment by paulscottrobson** [comment]
They actually might be but they're dynamically allocated as far as I can see.

**Comment by mist64** [comment]
A lot of locations have changed in the repo after r33, not the SYS ones though. They are a part of the $0300 vectors and will not change.

The SYS syntax with extra arguments will make it impossible to write helper tools callable using SYS that parse their own arguments, so I'm not a fan of that.

Close?

**Comment by greg-king5** [comment]
Utilities that have their own parameters simply could skip the four register arguments, then parse their things:
```bas
SYS PC,,,,, "FIRST UTILITY ARG", "SECOND ARG", "THIRD ARG"
```

**Comment by BruceMcF** [comment]
> Utilities that have their own parameters simply could skip the four register arguments, then parse their things:
> 
> ```vbnet
> SYS PC,,,,, "FIRST UTILITY ARG", "SECOND ARG", "THIRD ARG"
> ```

How exactly does the machine language routine know where those "utility arguments" are to be found, and how many to look for? A, X and Y would normally be used to pass utility arguments, or a pointer+len/count for their location, to the file.

The order of the register parameters then guide the writing of the ML program ... if there is one piece of information to pay, pass it in A if it is one byte, in A/X if it is 16bits, to pass two integers, have A refer to a ZP location (like Kernel SAVE) and X/Y hold the other one, if there are more than three arguments or a variable number, have A hold a count of how many arguments and X/Y point to the first item in a linked list.

**Comment by greg-king5** [comment]
> ```bas
> SYS PC,,,,, "FIRST UTILITY ARG", "SECOND ARG", "THIRD ARG"
> ```

The utility subroutine acts as a BASIC extension.  It uses the "wedge" method to hook into the interpreter.  It calls the `CHARGET` function (in the zero-page) to read the rest of the `SYS` statement.  On Commodore computers (where we know the contents of the ROM), it can call functions within the interpreter to help it parse that statement.

**Comment by kktos** [comment]
Just my 2cts:
what about having some variations of SYS like:
- **SYS**
_cbm basic v2_
SYS address
- **SYS7**
_cbm basic v7_
SYS7 address , < registers > [, < parameters >]
- **SYS7P**
SYS7P address , < parameters >

All this to compensate the fact we cannot name our parameters as with:
`SYS $3000, A:0, Y:$FF, "parm1"`

Or we can have something more C-like as a format string for printf.....
```
SYS "%PC%A%Y%P", $3000,255,0,"hello"
SYS "%PC%P", $3000,,"hello"
```
but I definitively prefer the former.

**Comment by mist64** [comment]
Th C64 design makes you POKE780+ with A/X/Y/P to run arbitrary machine code that is not aware of BASIC. To get the state back, PEEK780+. Pretty simple, super flexible. It also allows assembly helper code that is meant to be called only by BASIC that can use CHKCOM, GETBYT etc. to get extra arguments, in any syntax.

The proposed solutions (including the BASIC7 one) allow passing and retrieving A/X/Y/P with less syntax, but make the BASIC helper code case more verbose.

I'm still not convinced, but I am not saying I can't be convinced. :)

**Comment by greg-king5** [comment]
Yes, the Assembly code would be more verbose; but, so what?!  Assembly code is verbose anyway.  Adding some more `jsr` instructions to skip stuff that the helper doesn't need isn't a burden -- it's just a normal part of designing machine language subroutines.

The point is that the BASIC code would be _less_ verbose.  And, the special use of `POKE780+` and `PEEK(780+)` is harder to remember than the extended -- but simpler -- syntax of `SYS` and `RREG`.

**Comment by EgonOlsen71** [comment]
Is this actually a real world issue? During my C64 years, I never encountered this situation and I did quite a lot of BASIC/assembly mix code. I wasn't aware that this option actually exists. I only recently came across one program that used this and therefor, it failed to compile properly because I didn't support this A/X/Y with SYS in my compiler (because, as said, it never occured to me that this is supported by BASIC). None of my dozens of other BASIC test programs made use of this.
I think, there are more important things to tackle than making the SYS syntax more approachable. As long as the memory locations stay at a fixed position, I'm happy to poke into them (or peek from them) for the very few times that this is actually needed.

**Comment by BruceMcF** [comment]
> > ```vbnet
> > SYS PC,,,,, "FIRST UTILITY ARG", "SECOND ARG", "THIRD ARG"
> > ```
 
> The utility subroutine acts as a BASIC extension. It uses a variation of the "wedge" method to hook into the interpreter. It calls the `CHARGET` function (in the zero-page) to read the rest of the `SYS` statement. On Commodore computers (where we know the contents of the ROM), it can call functions within the interpreter to help it parse that statement.

So the issue is that unlike C128 SYS, C64 SYS takes no additional parameters, so ignores what you put after it, and then people have taken advantage of that to add utility arguments which they can use some of Basic's mechanisms to help parse?

I'm wondering if that's is really the cleanest way to do it, or whether it was simply the available way to do it with V2 Basic.

Turning ARG around to follow SYS, it would eliminate any conflict between parameters for Basic-helper routines and parameters for normal assembly language routines:

100 SYS $BFFE,15,%10000000

or:
100 LOOKUP = $BFF4
...
450 SYS LOOKUP :ARG"/[A-Z][a-z]*/*,$7F

**Comment by BruceMcF** [comment]
> A lot of locations have changed in the repo after r33, not the SYS ones though. They are a part of the $0300 vectors and will not change.
> 
> Close?

Although the side discussion would then be in the closed section, the main topic is resolved. This could be closed.

Regarding the side discussion:
> The SYS syntax with extra arguments will make it impossible to write helper tools callable using SYS that parse their own arguments, so I'm not a fan of that.

I would propose:
(1) SYS remain V2 SYS. If a RAMBANK specification keyword is put into place, it should be inherited by SYS, any call within $A000-$BFFF calling into the specified RAMBANK and any other routine having the specified RAMBANK available for data or support routines. This should also be understood to have the Basic ROM in scope, for the use of Basic Helper routines relying on SYS and it's flexibility in adding additional parameters after the address.

(2) The far call Kernel routine be exported to Basic as FARSYS, using V7 parameters:

FARSYS rombank, address [, A [, X [, Y [, P]]]]
